### ISSUE TITLE
Many timeouts on SOA queries while checking slave zones on startup

### ISSUE THREAD

**Issue by paddg** [PowerDNS/pdns#9112]
- Program: Authoritative
 - Issue type: Bug report

### Short description
Each time, we start pdns_server, it complaints because of  timeouts on SOA queries while checking slave zones:
`Received serial number updates for 1873 zones, had 127 timeouts`

### Environment
 - Operating system: SLES15-SP1
 - Software version: PowerDNS Authoritative Server 4.3.0
 - Software source: self compiled

### Steps to reproduce

My lab setup consists of two servers:
node22 (10.10.10.22)
node23 (10.10.10.23)

#### Configuration node22:
`# cat /etc/pdns/pdns.conf`
```
local-address=10.10.10.22
master=yes
launch=bind
bind-config=/etc/powerdns/bind/named.conf
allow-axfr-ips=10.10.10.23
```
`# head -10 /etc/powerdns/bind/named.conf` # (2000 zones)
```
options {
   directory "/var/powerdns/bind/";
};

zone "1.test." { type master; file "testzone"; };
zone "2.test." { type master; file "testzone"; };
zone "3.test." { type master; file "testzone"; };
zone "4.test." { type master; file "testzone"; };
zone "5.test." { type master; file "testzone"; };
zone "6.test." { type master; file "testzone"; };
```
`# cat /var/powerdns/bind/testzone `
```
@ 10 IN SOA node22. hostmaster.testzone. (1 10000 1800 3600000 3600)
@ 10 IN NS  node23.
```
#### Configuration node23:
`# cat /etc/pdns/pdns.conf`
```
local-address=10.10.10.23
slave=yes
launch=bind
bind-config=/etc/powerdns/bind/named.conf
```
`# head -10 /etc/powerdns/bind/named.conf` # (2000 zones)
```
options {
   directory "/var/powerdns/bind/";
};

zone "1.test." { type slave; file "1.test"; masters { 10.10.10.22; };};
zone "2.test." { type slave; file "2.test"; masters { 10.10.10.22; };};
zone "3.test." { type slave; file "3.test"; masters { 10.10.10.22; };};
zone "4.test." { type slave; file "4.test"; masters { 10.10.10.22; };};
zone "5.test." { type slave; file "5.test"; masters { 10.10.10.22; };};
zone "6.test." { type slave; file "6.test"; masters { 10.10.10.22; };};
```
Start pdns_server on node22:
`# pdns_server --guardian=no --daemon=no --disable-syslog --log-timestamp=no --write-pid=no`

Start pdns_server on node23:
`# pdns_server --guardian=no --daemon=no --disable-syslog --log-timestamp=no --write-pid=no 2>&1 | tee /var/log/pdns_server.output`

In my tests, each time I start pdns_server on node23, I get this timeouts message:
`Received serial number updates for 1873 zones, had 127 timeouts`
The numbers differs.

### Expected behaviour

It should be possible to check the SOA RRs on startup without getting timeouts.

### Actual behaviour
Many timeouts on SOA queries while checking slave zones on startup.

**Comment by Habbie** [comment]
Doing this test with 2 instances on localhost, running from git master or 4.3.0, I have no trouble getting all 2000 queried successfully in a fraction of a second. Do you think you could make some pcaps on both ends?

**Comment by paddg** [comment]
Hello @Habbie, I checked the pcaps. There are all 2000 queries and replies in it on both nodes. Within 58.921ms (34kqps). And I saw that the exact number of the timeout replies was dropped by the kernel:
`Received serial number updates for 1915 zones, had `**85**` timeouts`
**Before:**
```
node23:~ # netstat -s | grep err
    1185 packet receive errors
    1185 receive buffer errors
```
**After:**
```
node23:~ # netstat -s | grep err
    1270 packet receive errors
    1270 receive buffer errors
```
`1270-1185=`**85**

We did no IP stack/kernel/nic tuning on this lab server and it is a bit slow (2.5Ghz), but I think it should still work.

**Comment by phonedph1** [comment]
> We did no IP stack/kernel/nic tuning on this lab server and it is a bit slow (2.5Ghz), but I think it should still work.

Well `net.core.rmem_default` and `net.core.rmem_max`  are normally quite small, so small that it's not reasonable to expect them not to fill under bursts of traffic that an application may not be able to process as fast as you'd like.

**Comment by paddg** [comment]
Yes. But also, pdns_server should not send out SOA queries faster than it is able to receive the answers

**Comment by rgacogne** [comment]
The current code already limits the number of "in-flight" queries to 200 (`d_maxInFlight`, set from `CommunicatorClass::slaveRefresh`), sending at most two queries (`d_burst`, default value) at a time before looking if a response is available, to limit the query rate growth. That does seem reasonable given the size of SOA answers, but I guess we might want to make the number of in-flight queries configurable.
The current code also does favor the sending of queries over the receiving of answers, as long as we are below the maximum number of in-flight queries, which is something we might also consider changing (processing all received answers instead of trying to send a new query as soon as we have processed one answer).

**Comment by rgacogne** [comment]
> The current code also does favor the sending of queries over the receiving of answers, as long as we are below the maximum number of in-flight queries, which is something we might also consider changing (processing all received answers instead of trying to send a new query as soon as we have processed one answer).

Note that it would mean changing the way we currently use `poll()` in a blocking way, we should probably not block until we don't have enough room to send more queries.
